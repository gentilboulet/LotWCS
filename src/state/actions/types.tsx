import { IInitialStateAction } from './initial';
import { IHeaderAction } from './header';
import { IChiAction } from './chi';
import { ISkillAction } from './skills';
import { ILoresheetAction } from './loresheets';
import { IHistoryAction } from './history';
import { IKungFuAction } from './kungfus';
import { IVirtueAction } from './virtues';

export type IAction =
  IInitialStateAction | IHeaderAction | ISkillAction | IHistoryAction |
  ILoresheetAction | IChiAction | IKungFuAction | IVirtueAction;
