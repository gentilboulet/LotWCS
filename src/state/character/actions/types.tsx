import { ActionType } from "typesafe-actions";

import * as chi from "./chi";
import * as header from "./header";
import * as kungfu from "./kungfu";
import * as loresheets from "./loresheets";
import * as skills from "./skills";
import * as virtues from "./virtues";

export type ICharacterAction =
  | ActionType<typeof chi>
  | ActionType<typeof header>
  | ActionType<typeof kungfu>
  | ActionType<typeof loresheets>
  | ActionType<typeof skills>
  | ActionType<typeof virtues>;
