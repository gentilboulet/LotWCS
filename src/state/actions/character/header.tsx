import { createAction } from "typesafe-actions";

import * as dataArchetypes from "../../../data/archetypes";
import * as dataRanks from "../../../data/ranks";
import { IPerk } from "../../../perks";

export const setName = createAction("header/SET_NAME", (name: string) => {
  return { name };
})();

export const setConcept = createAction(
  "header/SET_CONCEPT",
  (concept: string) => {
    return { concept };
  },
)();

export const setArchetype = createAction(
  "header/SET_ARCHETYPE",
  (archetype: dataArchetypes.TArchetype) => {
    dataArchetypes.validateArchetype(archetype);
    return { archetype };
  },
)();

export const setRank = createAction(
  "header/SET_RANK",
  (rank: dataRanks.TRank) => {
    dataRanks.validateRank(rank);
    const dr = dataRanks.getRank(rank) as dataRanks.IDataRank;
    return { rank, perks: dr.perks as IPerk[] };
  },
)();
