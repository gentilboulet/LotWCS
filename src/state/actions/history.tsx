import * as constants from 'state/constants/history';

export interface IhistoryDeleteUpTo {
  type: constants.HISTORY_DELETE;
  id: number;
}

export type IHistoryAction = IhistoryDeleteUpTo;

export function historyDeleteUpTo(id: number): IhistoryDeleteUpTo {
  return { type: constants.HISTORY_DELETE, id };
}
