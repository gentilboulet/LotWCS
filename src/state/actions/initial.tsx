import * as constants from 'constants/initial';

export interface IInitialState {
  type: constants.INITIAL_STATE;
}

export type IInitialStateAction = IInitialState;

export function resetToInitialState(): IInitialState {
  return {type: constants.INITIAL_STATE};
}
