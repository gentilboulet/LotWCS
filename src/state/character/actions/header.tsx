import { createAction } from "typesafe-actions";

import * as dataArchetypes from "../../../data/archetypes";
import * as dataRanks from "../../../data/ranks";

export const setName = createAction(
  "header/SET_NAME",
  action => (name: string) => {
    return action({ name });
  },
);

export const setConcept = createAction(
  "header/SET_CONCEPT",
  action => (concept: string) => {
    return action({ concept });
  },
);

export const setArchetype = createAction(
  "header/SET_ARCHETYPE",
  action => (archetype: dataArchetypes.TArchetype) => {
    dataArchetypes.validateArchetype(archetype);
    return action({ archetype });
  },
);

export const setRank = createAction(
  "header/SET_RANK",
  action => (rank: dataRanks.TRank) => {
    dataRanks.validateRank(rank);
    return action({ rank });
  },
);
