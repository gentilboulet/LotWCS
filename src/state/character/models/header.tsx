import * as data from "../../../data/ranks";
import { ICharacterState } from "./type";

export function setRank(state: ICharacterState, rank: data.TRank): void {
  const dataRank = data.getRank(rank) as data.IDataRank;
  state.rank = dataRank.value;
}
