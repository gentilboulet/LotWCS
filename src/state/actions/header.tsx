import * as constants from 'state/constants/header';

import * as dataArchetypes from 'data/archetypes';
import * as dataRanks from 'data/ranks';

export interface IHeaderSetName {
  type: constants.HEADER_SET_NAME;
  name: string;
}

export interface IHeaderSetConcept {
  type: constants.HEADER_SET_CONCEPT;
  concept: string;
}

export interface IHeaderSetArchetype {
  type: constants.HEADER_SET_ARCHETYPE;
  archetype: string;
}

export interface IHeaderSetRank {
  type: constants.HEADER_SET_RANK;
  rank: string;
}

export type IHeaderAction = IHeaderSetName | IHeaderSetConcept | IHeaderSetArchetype | IHeaderSetRank;

export function setName(name: string): IHeaderSetName {
  return { type: constants.HEADER_SET_NAME, name };
}

export function setConcept(concept: string): IHeaderSetConcept {
  return { type: constants.HEADER_SET_CONCEPT, concept };
}

export function setArchetype(archetype: string): IHeaderSetArchetype {
  dataArchetypes.validateArchetype(archetype);
  return { archetype, type: constants.HEADER_SET_ARCHETYPE };
}

export function setRank(rank: string): IHeaderSetRank {
  dataRanks.validateRank(rank);
  return { rank, type: constants.HEADER_SET_RANK };
}
