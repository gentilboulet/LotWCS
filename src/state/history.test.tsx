import { IAction } from 'state/actions/types';
import { IStoreState } from 'state/types';

import * as header from 'state/actions/header';
import { resetToInitialState } from 'state/actions/initial';
import { initialStateFactory } from 'state/initial';

import { pushToHistory, replayHistory } from './history';

const initialState: IStoreState = initialStateFactory();

describe('Testing pushToHistory', () => {
  it('should push actions to history', () => {
    const action = header.setName('Dummy Name');
    expect( initialState.getIn(['history', 0]) ).toMatchObject( resetToInitialState() );

    const state = pushToHistory(initialState, action);
    expect( state.getIn(['history', 0]) ).toMatchObject( initialState.getIn(['history', 0]) );
    expect( state.getIn(['history', 1]) ).toMatchObject( action );
  });
});

describe('Testing replayHistory', () => {
  it('should replay an history of actions', () => {
    const actions: IAction[] = [
      resetToInitialState(),
      header.setName('Roberts'),
      header.setConcept('Dread pirate Robert'),
      resetToInitialState(),
      header.setName('John')
    ];
    const state = replayHistory(initialState, actions);
    expect( state ).toMatchSnapshot();
    expect( state.get('name') ).toBe('John');
    expect( state.get('concept') ).toBe('No Concept');
    expect( state.get('history').length === (actions.length + 1) );
  });
});
