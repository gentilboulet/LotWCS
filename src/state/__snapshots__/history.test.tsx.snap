// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing replayHistory should replay an history of actions 1`] = `
Object {
  "archetype": undefined,
  "bonuses": Array [],
  "chi": Object {
    "corrupt": Object {
      "cultivation": 0,
      "value": 0,
    },
    "earth": Object {
      "cultivation": 0,
      "value": 0,
    },
    "enlightened": Object {
      "cultivation": 0,
      "value": 0,
    },
    "fire": Object {
      "cultivation": 0,
      "value": 0,
    },
    "general": Object {
      "cultivation": 0,
      "value": 0,
    },
    "metal": Object {
      "cultivation": 0,
      "value": 0,
    },
    "water": Object {
      "cultivation": 0,
      "value": 0,
    },
    "wood": Object {
      "cultivation": 0,
      "value": 0,
    },
  },
  "concept": undefined,
  "destiny": 0,
  "discounts": Array [],
  "entanglement": 0,
  "history": Array [
    Object {
      "type": "INITIAL_STATE",
    },
    Object {
      "name": "John",
      "type": "HEADER_SET_NAME",
    },
  ],
  "kungfu": Object {
    "KUNGFU_EXTERNAL": Object {},
    "KUNGFU_INTERNAL": Object {},
  },
  "loresheets": Object {},
  "name": "John",
  "rank": undefined,
  "skills": Object {
    "Awareness": Object {
      "name": "Awareness",
      "specialities": Array [],
      "value": 0,
    },
    "Confidence": Object {
      "name": "Confidence",
      "specialities": Array [],
      "value": 0,
    },
    "Crafting": Object {
      "name": "Crafting",
      "specialities": Array [],
      "value": 0,
    },
    "Finesse": Object {
      "name": "Finesse",
      "specialities": Array [],
      "value": 0,
    },
    "Hardiness": Object {
      "name": "Hardiness",
      "specialities": Array [],
      "value": 0,
    },
    "Inspire": Object {
      "name": "Inspire",
      "specialities": Array [],
      "value": 0,
    },
    "Learning": Object {
      "name": "Learning",
      "specialities": Array [],
      "value": 0,
    },
    "Medicine": Object {
      "name": "Medicine",
      "specialities": Array [],
      "value": 0,
    },
    "Might": Object {
      "name": "Might",
      "specialities": Array [],
      "value": 0,
    },
    "Perform": Object {
      "name": "Perform",
      "specialities": Array [],
      "value": 0,
    },
    "Politics": Object {
      "name": "Politics",
      "specialities": Array [],
      "value": 0,
    },
    "Ride": Object {
      "name": "Ride",
      "specialities": Array [],
      "value": 0,
    },
    "Stealth": Object {
      "name": "Stealth",
      "specialities": Array [],
      "value": 0,
    },
    "Survival": Object {
      "name": "Survival",
      "specialities": Array [],
      "value": 0,
    },
    "Tactics": Object {
      "name": "Tactics",
      "specialities": Array [],
      "value": 0,
    },
    "Wu Wei": Object {
      "name": "Wu Wei",
      "specialities": Array [],
      "value": 0,
    },
  },
  "virtues": Array [
    Object {
      "name": "Honor",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Benevolence",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Righteousness",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Loyalty",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Force",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Revenge",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Object {
      "name": "Individualism",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Object {
      "name": "Obsession",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Object {
      "name": "Ruthlessness",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Object {
      "name": "Ferocity",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
  ],
}
`;
